security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,428 @@
+
+      /**
+       * Handle value where the persistent key is expected to be found
+       */
+      public handle: TPM_HANDLE,
+
+      /**
+       * Template to be used for key cretaion if the persistent key with the given handle does not exist
+       */
+      public template: TPMT_PUBLIC
+
+  ) {}
+}
+
 
 export class TpmSecurityClient  {
-  getEndorsementKey(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+
+  private _ek: TPMT_PUBLIC = null;
+  private _srk: TPMT_PUBLIC = null;
+  private _registrationId: string = '';
+  private _tpm: Tpm;
+  private _fsm: machina.Fsm;
+  private _idKeyPub: TPMT_PUBLIC = null;
+
+
+  constructor(useSimulator: boolean, registrationId?: string) {
+    this._tpm = new Tpm(useSimulator);
+    if (registrationId) {
+      this._registrationId = registrationId;
+    }
+    this._fsm = new machina.Fsm({
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (callback, err) => {
+            if (callback) {
+              if (err) {
+                callback(err);
+              } else {
+                callback(null, null);
+              }
+            }
+          },
+          connect: (connectCallback) => this._fsm.transition('connecting', connectCallback),
+          getEndorsementKey: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('getEndorsementKey', callback);
+              }
+            });
+          },
+          getStorageRootKey: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('getStorageRootKey', callback);
+              }
+            });
+          },
+          signWithIdentity: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('signWithIdentity', callback);
+              }
+            });
+          },
+          activateSymmetricIdentityIdentity: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('activateSymmetricIdentityIdentity', callback);
+              }
+            });
+          },
+          disconnect: (callback) => {
+            if (callback) {
+              callback();
+            }
+           }
+        },
+        connecting: {
+          _onEnter: (callback, err) => {
+            try {
+              let self = this;
+              this._tpm.connect(function (): void {
+                self._createPersistentPrimary(new PersKeyInfo('EK', tss.Endorsement, ekPersHandle, ekTemplate), (ekCreateErr, ekPublicKey) => {
```

Comment:
>self._createPersistentPrimary [](start = 16, length = 29)

why do this here and not in the calls to getEndorsementKey and getStorageRootKey?

Reply:
As discussed.  Making the ek and the srk persistent tpm objects is the essential first step in interacting with the tpm.  This step is bascially why we are using an FSM.

---
In reply to: [157888474](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r157888474) [](ancestors = 157888474)